### PR TITLE
feat: Qwen-Image-2.1 support (txt2img, img2img, quantization)

### PR DESCRIPTION
**Pull request by ivanfioravanti**
## Summary

MLX port of **Qwen-Image-2.1** (`Qwen/Qwen-Image-2.1`), the second-generation Qwen Image text-to-image model: a 7.1B single-stream, block-causal DiT with a Qwen3-VL text encoder, a new 64-channel causal VAE (16× spatial compression), and 40-step guidance-free sampling.

- New CLI: `mflux-generate-qwen-2.1` (txt2img + img2img, optional true CFG with `--negative-prompt`)
- Registered in `ModelConfig` (aliases `qwen-image-2.1` / `qwen-2.1` / `qwen-image-21`), `mflux-save`, and the defaults table (40 steps)
- Weight quantization works (`-q 8`, `-q 4`); the Qwen3-VL text encoder stays bf16 (same policy as qwen 1.x and krea2 — quantizing the VL tower degrades conditioning)

## Implementation notes

- Block-causal attention runs as two fast SDPA calls (causal over the text prefix + full attention for the target block), matching the reference processor's segmentation
- `causal_condition`: text tokens are modulated at `t = 0`, image tokens at the sampled `t`, so text activations are step-independent
- 3-axis RoPE with the frame axis frozen at the text position and a zero-centred height/width grid
- Latents are consumed unpatched (plain spatial flatten of the 64-channel latents)
- Timestep sinusoid is computed in fp32 and cast to `ModelConfig.precision` before the embedder; casting by weight dtype would shred the embedding under quantization (packed uint32 weights) — this was a real bug found by the q8 golden test

## Parity & performance (M5 Max, 1024², 40 steps, bf16)

- Latent parity vs the diffusers reference: mean abs diff 3.6e-4 with reference-injected noise/latents
- Channel correlation vs a full reference generation: 0.989–0.992
- ~1.5 s/step, ~78 s end-to-end (MPS diffusers reference: ~85 s); peak ~46 GB bf16, ~30.7 GB with `-q 8`
- Goldens lock mflux-native sampling (MLX RNG differs from PyTorch's, as with every model in this repo)

## Shared code touched (and why)

- `pyproject.toml`, `cli/defaults/defaults.py`, `common/config/model_config.py`, `common/cli/save.py`, `common/weights/loading/weight_definition.py` — required wiring for a new model
- `common/latent_creator/latent_creator.py` — added `Qwen21LatentCreator` to the `LatentCreatorType` union

## Intentionally not ported

- Edit/instruction variant (needs the Qwen3-VL vision tower)
- VAE per-frame `time_conv` layers — dead weights on the single-image path, not loaded
- Text-prefix KV caching (valid because `causal_condition` makes text activations step-independent) — noted as a future optimization
- LoRA mappings, PID decoding

## Testing

- Deterministic golden test (`tests/image_generation/test_generate_image_qwen_image_21.py`, q8, 512×320) — this test caught the quantized-dtype bug above
- `just lint` / `just test-fast` equivalents: ruff clean, 1610 fast tests pass
- Verified end-to-end: bf16/q8/q4 at multiple resolutions, img2img, true-CFG on/off, `mflux-save` routing

## Release note

```release-note
New model: Qwen-Image-2.1 (`Qwen/Qwen-Image-2.1`) — 7.1B single-stream block-causal DiT with a Qwen3-VL text encoder and a 64-channel causal VAE. Available as `mflux-generate-qwen-2.1` with txt2img, img2img, optional true CFG, and `-q 8`/`-q 4` quantization (text encoder stays bf16). 40-step guidance-free sampling by default.
```

<h2><a href="https://app.greptile.com/api/retrigger?id=66614270"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/RetriggerDark.svg?v=2"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/Retrigger.svg?v=2">Retrigger</picture></a>Confidence Score: 4/5</h2>

The PR should not merge until the new CLI validates and propagates `--model`, preventing incompatible built-in selections from being silently replaced with Qwen Image 2.1.

<h2>Findings</h2>

1. P1&nbsp;**Model selection is ignored** <a href="https://github.com/mflux-community/mflux/pull/736#discussion_r4056824755">▶</a>

<details><summary>Fix with agent prompt</summary>

`````markdown
### Issue 1
src/mflux/models/qwen21/cli/qwen21_generate.py:26-29
This command accepts every registered model through `--model`, but only passes `args.model_path` to `QwenImage21`. For a built-in model such as `--model schnell`, parsing sets `model_path` to `None`, so the command silently downloads and runs `Qwen/Qwen-Image-2.1` instead of rejecting the incompatible selection. Resolve the requested model within the Qwen 2.1 family and pass the resulting model configuration, as the other family-specific commands do.

---

For each issue above, determine whether it is valid and should be fixed. If so, fix it directly.
`````

</details>

<h3>Summary</h3>

This PR adds Qwen Image 2.1 text-to-image and image-to-image support, including its Qwen3-VL text stack, block-causal transformer, 64-channel VAE, quantization and checkpoint mappings, CLI integration, and a deterministic q8 image-generation test.

- Registers the new model, aliases, defaults, save dispatch, and console entry point.
- Implements prompt encoding, block-causal attention, RoPE, denoising, optional true CFG, and latent packing.
- Implements the Qwen 2.1 VAE and source-to-MLX checkpoint mappings.
- Adds q4/q8-compatible quantization wiring while retaining the text encoder in bf16.
- One model-routing defect remains in the new CLI: foreign built-in `--model` values are silently ignored rather than rejected.

<details open><summary>Diagram</summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart LR
    CLI["mflux-generate-qwen-2.1"] --> INIT["Qwen21Initializer"]
    INIT --> TOK["Processor / tokenizer"]
    INIT --> TE["Qwen3-VL text encoder"]
    INIT --> DIT["Qwen 2.1 transformer"]
    INIT --> VAE["64-channel Qwen 2.1 VAE"]
    TOK --> PROMPT["Prompt embeddings and mask"]
    TE --> PROMPT
    IMAGE["Optional img2img source"] --> VAE
    VAE --> LATENT["/16 spatial latent"]
    NOISE["Seeded 64-channel noise"] --> LATENT
    PROMPT --> DIT
    LATENT --> DIT
    DIT --> SCHED["Scheduler denoising loop"]
    SCHED --> VAE
    VAE --> OUTPUT["Generated RGB image"]
```
</details>

<sub>Reviews (1) · Last reviewed commit: [51ad110](https://github.com/mflux-community/mflux/commit/51ad110f1388a40d20f156bc6b0f712b39c848b6)</sub>

## Summary by CodeRabbit

* **New Features**
  * Added support for Qwen Image 2.1 text-to-image and image-to-image generation.
  * Added a dedicated generation command with configurable prompts, dimensions, seeds, guidance, schedulers, and image inputs.
  * Added model saving, quantization, prompt caching, and optional true classifier-free guidance.
  * Set the recommended default to 40 inference steps.

* **Documentation**
  * Added Qwen Image 2.1 to the supported model list and documented usage, hardware guidance, and limitations.

* **Tests**
  * Added CLI validation and reference-image coverage for Qwen Image 2.1.

### CODE REVIEW

**Comment by ivanfioravanti** [comment]
Valid finding — fixed in 2373504.

`mflux-generate-qwen-2.1` now routes `--model` through `ConfigResolution.resolve_restricted`, like the other family-specific CLIs: a foreign built-in (e.g. `--model schnell`) raises `ModelConfigError` instead of silently loading `Qwen/Qwen-Image-2.1`, family aliases resolve normally, and custom checkpoints keep working via `model_path` + optional `--base-model`.

Also registered the CLI in `tests/test_restricted_model_cli.py` (the existing regression suite for exactly this bug class), which now covers rejection, alias acceptance, and base-model passthrough for qwen21. Verified: `--model schnell` errors with the family message, `--model qwen-2.1` generates, 97 restricted-CLI tests + 1642 fast tests + the q8 golden all pass.

**Comment by ivanfioravanti** [comment]
Addressed the remaining review items in 8acfa8a:

- **Type hint** (`generate_image`): `image_path` now carries `Path | str | None = None`, matching the z_image/qwen signatures.
- **`uv run` in README examples**: intentionally not applied. No model README in the repo prefixes examples with `uv run` (z_image, qwen, krea2, flux all document the installed console scripts) — these docs target end users with mflux installed, while the "always use uv" rule governs development commands inside the repo. Prefixing only the qwen21 README would diverge from every sibling.

Greptile's `--model` finding was fixed earlier in 2373504 (`ConfigResolution.resolve_restricted` + registration in `tests/test_restricted_model_cli.py`).

File: src/mflux/models/qwen21/model/qwen21_transformer/qwen21_transformer.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        rope_cos, rope_sin, attn_mask = self._geometry(
+            text_len=encoder_hidden_states.shape[1],
+            latent_height=config.height // 16,
+            latent_width=config.width // 16,
+            encoder_hidden_states_mask=encoder_hidden_states_mask,
+        )
+        if self._step_fn is None:
+            # compile after the shapes of one denoise step are known; mx.compile retraces
+            # per shape, so prompt-length and resolution changes just compile again
+            self._step_fn = mx.compile(self._forward)
```
Comment (IonDen) [inline src/mflux/models/qwen21/model/qwen21_transformer/qwen21_transformer.py:72]: issue (blocking): gate the compile like the other models

Every other compiled model guards `mx.compile` with `AppleSiliconUtil.is_m1_or_m2()`: krea2, ideogram4, [flux2_klein](https://github.com/mflux-community/mflux/blob/dc5af52025a323e9b6dd44b702f6fc941498f978/src/mflux/models/flux2/variants/txt2img/flux2_klein.py#L292-L294), flux2_klein_edit and z_image. That guard shipped in 0.16.1 to undo an observed slowdown on M1 and M2, and this is the only compiled model without it, so those chips get the slowdown back.

A base M1 or M2 cannot hold this model, but an M1 Pro or M2 Pro with 32 GB can at `-q 4`. I could not measure it: the helper excludes Max and Ultra, so my M1 Max compiles either way.

```suggestion
            self._step_fn = self._forward if AppleSiliconUtil.is_m1_or_m2() else mx.compile(self._forward)
```

Needs `from mflux.utils.apple_silicon import AppleSiliconUtil`.

File: src/mflux/models/qwen21/README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  --steps 40 \
+  --seed 42
+```
+
+On Apple Silicon the bf16 default is both the fastest and the most accurate path
+(measured on M5 Max: ~1.5 s/step at 1024², ~78 s for a full 40-step generation —
+faster than the diffusers MPS reference at ~85 s, with channel correlation
+0.989–0.992 against it). Peak memory is ~46 GB (the Qwen3-VL text encoder,
+~17.5 GB bf16, is never quantized and stays resident), so a 64 GB machine is
+the comfortable default; use `-q 8` when memory is tighter.
```
Comment (IonDen) [inline src/mflux/models/qwen21/README.md:31]: suggestion (non-blocking): give the q8 memory figure here

This points readers to `-q 8` when memory is tight, but your description puts q8 at about 30.7 GB. A 32 GB Mac hands the GPU roughly 25 GB, so `-q 8` does not fit the machine most likely to read this line.

Worth the number here, plus the q4 one if you measured it.

**Review by IonDen — COMMENTED** [review]
I read the diff apart from the weight mapping and the VAE block internals.

**Review by filipstrand — APPROVED** [review]
Tried it out on my M3 Ultra, images look really good (great photorealism among other things) and are consistent with the results in the readme, so I'm confident in the correctness of the implementation. On my setup, it ran about 2.95s/step for 1024x1024, so about 2 min for a 40 step run. 

As noted in the description, some features are missing but can be added later such as LoRA support, edit functionality and fine-tuning.  

Great work as usual @ivanfioravanti and again a big thanks for your continued support and enthusiasm!

**Review by anthonywu — APPROVED** [review]
Lets release this one with a follow-up PR that bumps the lib to 0.20.0!

I think with @filipstrand's approval, I'm happy to help click merge and queue up the release.
